src/actions/auth/session.js

```diff
@@ -0,0 +1,66 @@
+'use strict';
+
+const Future = require('fluture');
+const permissions = require('config').get('permissions');
+const {either, concat, pipe, get, maybe} = require('sanctuary-env');
+const {objOf, map, chain} = require('ramda');
+const {ObjectId} = require('mongodb');
+const {maybeToFuture} = require('../../util/future');
+const error = require('http-errors');
+const mm = require('micromatch');
+const {getTokenFromRequest, tokenToSession} = require('./_util');
+
+//    authorizedGroups :: Array String
+const authorizedGroups = ['@authorized'];
+
+//    unauthorizedGroups :: Array String
+const unauthorizedGroups = ['@unauthorized'];
+
+//    userNotFound :: NotAuthorizedError
+const userNotFound = error(403, 'User provided by token does not exist');
+
+//    missingPermission :: String -> NotAuthorizedError
+const missingPermission = x => error(403, `You are missing the ${x} permission`);
+
+//    getUserGroups :: User -> Array Group
+const getUserGroups = pipe([
+  get(Array, 'groups'),
+  maybe(authorizedGroups, concat(authorizedGroups))
+]);
+
+//    groupsToPermissions :: Array Group -> Array String
+const groupsToPermissions = chain(group => permissions[group] || []);
+
+//Export a middleware which determines the user session and attaches it to request.auth.
+module.exports = req => {
+
+  //    findUserById :: UserId -> Future NotFoundError User
+  const findUserById = pipe([
+    ObjectId,
+    objOf('_id'),
+    req.services.users.get,
+    chain(maybeToFuture(userNotFound))
+  ]);
```

Comment:
I'm not sure if I should find the user here. It defeats one of the benefits of using JWT; "not having to contact the user server". Perhaps I should instead determine the permissions of a user when the token is created, and store it in the token itself. Since tokens have a lifetime of 6 hours, changing a users permissions will take effect after a maximum of six hours.